File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.*;

class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        while(true) {
            String s = br.readLine();
            if(s.equals(".")) {
                br.close();
                break;
            }
            sb.append(Balance(s)? "yes\n" : "no\n");
        }
        System.out.println(sb);
    }
    private static boolean Balance(String s){
        Stack<Character> stack = new Stack<>();

        for(char c : s.toCharArray()) {
            if(c == '(' || c == '[') stack.push(c);
            else if(c == ')' || c == ']') {
                if(stack.isEmpty()) return false;
                char top = stack.pop();
                if(c == ')' && top != '(' || c == ']' && top != '[') return false;
            }
        }
        return stack.isEmpty();
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
from sys import stdin
from collections import deque

while True:
    s = stdin.readline().rstrip()
    if s == '.':
        break
    
    stack = deque()
    error = False
    
    for char in s:
        if char in '([':
            stack.append(char)
        elif char == ']':
            if not stack or stack[-1] != '[':
                error = True
                break
            stack.pop()
        elif char == ')':
            if not stack or stack[-1] != '(':
                error = True
                break
            stack.pop()
            
    if stack: error = True
    print("no" if error else "yes")